src/java.base/share/classes/jdk/internal/foreign/abi/Specializer.java

```diff
@@ -0,0 +1,854 @@
+        // load all the leaf args
+        for (int i = 0; i < leafArgSlots.length; i++) {
+            emitLoad(leafArgTypes.get(i), leafArgSlots[i]);
+        }
+        // call leaf MH
+        mv.visitMethodInsn(INVOKEVIRTUAL, METHOD_HANDLE_INTRN, INVOKE_EXACT_NAME, leafType.descriptorString(), false);
+
+        // for downcalls, store the result of the leaf handle call away, until
+        // it is requested by a VM_LOAD in the return recipe.
+        if (callingSequence.forDowncall() && leafType.returnType() != void.class) {
+            emitSaveReturnValue(leafType.returnType());
+        }
+        // for upcalls we leave the return value on the stack to be picked up
+        // as an input of the return recipe.
+
+        // return value processing
+        if (callingSequence.hasReturnBindings()) {
+            if (callingSequence.forUpcall()) {
+                typeStack.push(leafType.returnType());
+            }
+
+            retBufOffset = 0; // offset for reading from return buffer
+            doBindings(callingSequence.returnBindings());
+
+            if (callingSequence.forUpcall() && !callingSequence.needsReturnBuffer()) {
+                // was VM_STOREd somewhere in the bindings
+                emitRestoreReturnValue(callerMethodType.returnType());
+            }
+            mv.visitLabel(tryEnd);
+            // finally
+            emitCloseContext();
+
+            if (callerMethodType.returnType() == void.class) {
+                // The case for upcalls that return by return buffer
+                assert typeStack.isEmpty();
+                mv.visitInsn(RETURN);
+            } else {
+                assert typeStack.pop() == callerMethodType.returnType();
+                assert typeStack.isEmpty();
+                emitReturn(callerMethodType.returnType());
+            }
+        } else {
+            assert callerMethodType.returnType() == void.class;
+            assert typeStack.isEmpty();
+            mv.visitLabel(tryEnd);
+            // finally
+            emitCloseContext();
+            mv.visitInsn(RETURN);
+        }
+
+        mv.visitLabel(catchStart);
+        // finally
+        emitCloseContext();
+        if (callingSequence.forDowncall()) {
+            mv.visitInsn(ATHROW);
+        } else {
+           mv.visitMethodInsn(INVOKESTATIC, SHARED_UTILS_INTRN, HANDLE_UNCAUGHT_EXCEPTION_NAME, HANDLE_UNCAUGHT_EXCEPTION_DESC, false);
+           if (callerMethodType.returnType() != void.class) {
+               emitConstZero(callerMethodType.returnType());
+               emitReturn(callerMethodType.returnType());
+           } else {
+               mv.visitInsn(RETURN);
+           }
+        }
+
+        mv.visitTryCatchBlock(tryStart, tryEnd, catchStart, null);
+    }
+
+    private void doBindings(List<Binding> bindings) {
+        for (Binding binding : bindings) {
+            switch (binding.tag()) {
+                case VM_STORE -> emitVMStore((Binding.VMStore) binding);
```

Comment:
I note that we depart from the previous model of having virtual calls on the Binding instance which does the right thing. That's fine, I guess there's a lot of coupling (esp. in VMLoad/VMStore) with the specializer state.